pkg/bucketrequest/bucketrequest.go

```diff
@@ -0,0 +1,174 @@
+	"github.com/kubernetes-sigs/container-object-storage-interface-controller/pkg/util"
+	kubeclientset "k8s.io/client-go/kubernetes"
+
+	"github.com/golang/glog"
+)
+
+type bucketRequestListener struct {
+	kubeClient   kubeclientset.Interface
+	bucketClient bucketclientset.Interface
+}
+
+func NewListener() bucketcontroller.BucketRequestListener {
+	return &bucketRequestListener{}
+}
+
+func (b *bucketRequestListener) InitializeKubeClient(k kubeclientset.Interface) {
+	b.kubeClient = k
+}
+
+func (b *bucketRequestListener) InitializeBucketClient(bc bucketclientset.Interface) {
+	b.bucketClient = bc
+}
+
+// Add creates a bucket in response to a buckerrequest
+func (b *bucketRequestListener) Add(ctx context.Context, obj *v1alpha1.BucketRequest) error {
+	glog.V(1).Infof("add called for bucket %s", obj.Name)
+	bucketRequest := obj
+	status, err := b.provisionBucketRequestOperation(ctx, bucketRequest)
+	if err == nil || status == "Finished" {
+		// Provisioning is 100% finished / not in progress.
+		switch err {
+		case nil:
+			glog.V(5).Infof("BucketRequest processing succeeded, removing bucketRequest %s from bucketRequests in progress", bucketRequest.UID)
+		case util.ErrStopProvision:
+			glog.V(5).Infof("Stop provisioning, removing bucketRequest %s from bucketRequests in progress", bucketRequest.UID)
+			// Our caller would requeue if we pass on this special error; return nil instead.
+			err = nil
+		default:
+			glog.V(2).Infof("Final error received, removing buckerRequest %s from bucketRequests in progress", bucketRequest.UID)
+		}
+		return err
+	}
+	if status == "InBackground" {
+		//nothing for now
+	} else {
+		// status == ProvisioningNoChange.
+		// Don't change bucketRequestsInProgress:
+		// - the bucketRequest is already there if previous status was ProvisioningInBackground.
+		// - the bucketRequest is not there if if previous status was "Finished".
+	}
+	return nil
+}
+
+// update processes any updates  made to the bucket request
+func (b *bucketRequestListener) Update(ctx context.Context, old, new *v1alpha1.BucketRequest) error {
+	glog.V(1).Infof("update called for bucket %v", old)
+	return nil
+}
+
+// Delete processes a bucket for which bucket request is deleted
+func (b *bucketRequestListener) Delete(ctx context.Context, obj *v1alpha1.BucketRequest) error {
+	return nil
+}
+
+// provisionBucketRequestOperation attempts to provision a bucket for the given bucketRequest.
+// Returns nil error only when the bucket was provisioned (in which case it also returns "Finished"),
+// a normal error when the bucket was not provisioned and provisioning should be retried (requeue the bucketRequest),
+// or the special errStopProvision when provisioning was impossible and no further attempts to provision should be tried.
+func (b *bucketRequestListener) provisionBucketRequestOperation(ctx context.Context, bucketRequest *v1alpha1.BucketRequest) (string, error) {
```

Comment:
did you consider the func return being `(bool, error)` like is done for many of the _wait_ funcs?